### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by kaykay38**
All tests passed. Extra Credit attempted.

### CODE REVIEW

**Comment by worseTyler** [comment]
Assignment
Jester class takes in the correct interfaces: ✔
Jester dependencies are null checked: ✔
Jester appropriately retrieves jokes and sends them to the output: ✔
Chuck Norris jokes get filtered out: ✔
Jester code coverage is above 90%: ✔
Extra Credit
Output to console implementation is unit tested: ❌
JokeService uses JSON: ❌

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 1]
```diff
@@ -1,10 +1,86 @@
...
+            Jester jester = new Jester(new JokeService(), null!);
+            // Assert
+        }
+
+        [TestMethod]
+        public void Jester_ImplementIJokeServiceAndJokePrinter_WorkingJokeServiceAndJokePrinter()
+        {
+            // Arrange
+            IJokeService iJokeService = new JokeService();
+            IJokePrinter iJokePrinter = new JokePrinter();
```
Comment (Keboo) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:35]: In general we don't include the "i" in the variable name
```suggestion
            IJokeService jokeService = new JokeService();
            IJokePrinter jokePrinter = new JokePrinter();
```

File: CanHazFunny/CanHazFunny.Tests/JesterTests.cs [hunk 2]
```diff
@@ -1,10 +1,86 @@
...
+        public void TellJoke_GivenValidJoke_ExcecutesWithoutException()
+        {
+            //Arrange
+            string joke = "Some bad dad joke that makes you roll your eyes.";
+            Mock<IJokeService> mockJokeService = new Mock<IJokeService>();
+            mockJokeService.SetupSequence(JokeService => JokeService.GetJoke())
+                .Returns(joke);
+
+            Mock<IJokePrinter> mockJokePrinter = new Mock<IJokePrinter>();
+            mockJokePrinter.SetupSequence(JokePrinter => JokePrinter.PrintJoke(joke));
```
Comment (Keboo) [inline CanHazFunny/CanHazFunny.Tests/JesterTests.cs:55]: We will get to this later in the class when we cover lambdas, but we typically name the labdam argument with camelCase
```suggestion
            mockJokeService.SetupSequence(jokeService => jokeService.GetJoke())
                .Returns(joke);

            Mock<IJokePrinter> mockJokePrinter = new Mock<IJokePrinter>();
            mockJokePrinter.SetupSequence(jokePrinter => jokePrinter.PrintJoke(joke));
```

File: CanHazFunny/CanHazFunny.Tests/JokePrinterTests.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            // Arrange
+            string joke="Some bad dad joke that makes you roll your eyes.";
+            JokePrinter jokePrinter = new JokePrinter();
+
+            // Act
+            // Pipes Console.WriteLine to a file to test if the 
+            // JokePrinter.PrintJoke method prints the correct output to console.
+            using ( StreamWriter streamWriter= new StreamWriter("out.txt"))
+            {
+                Console.SetOut(streamWriter);
```
Comment (Keboo) [inline CanHazFunny/CanHazFunny.Tests/JokePrinterTests.cs:22]: :+1:

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+using System;
+
+namespace CanHazFunny
+{
+    public class Jester
+    {
+        public IJokePrinter jokePrinter {get; private set;}
+        public IJokeService jokeService {get; private set;}
```
Comment (Keboo) [inline CanHazFunny/CanHazFunny/Jester.cs:8]: Properties should be named with PascalCase
```suggestion
        public IJokePrinter JokePrinter {get; private set;}
        public IJokeService JokeService {get; private set;}
```

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+        public Jester(IJokeService jokeService, IJokePrinter jokePrinter)
+        {
+            this.jokeService = jokeService ?? throw new ArgumentNullException();
+            this.jokePrinter = jokePrinter ?? throw new ArgumentNullException();
```
Comment (Keboo) [inline CanHazFunny/CanHazFunny/Jester.cs:13]: When throwing ArgumentNullExceptions always specify the argument name
```suggestion
            this.jokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
            this.jokePrinter = jokePrinter ?? throw new ArgumentNullException(nameof(jokePrinter));
```

File: CanHazFunny/CanHazFunny/JokeService.cs [hunk 6]
```diff
@@ -2,13 +2,13 @@
...
 {
-    public class JokeService
+    public class JokeService : IJokeService
     {
         private HttpClient HttpClient { get; } = new();
 
         public string GetJoke()
         {
-            string joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api").Result;
+            string joke = HttpClient.GetStringAsync("https://geek-jokes.sameerkumar.website/api?format=json").Result;
```
Comment (Keboo) [inline CanHazFunny/CanHazFunny/JokeService.cs:11]: This is good, but doesn't actually parse parse the json.

**Review by Keboo — COMMENTED** [review]
Great work!

It looks like you deleted the Docs/Lectures folders. Normally this should not be needed as it just increased the number of modified files in your pull request.

**Pull request closed without merging**